From 4650b9717debe41b2114e8923053f7ccb5862d96 Mon Sep 17 00:00:00 2001
From: Bryan Cantrill <bryan@joyent.com>
Date: Tue, 19 Jul 2016 01:24:42 +0000
Subject: [PATCH] OS-5448 clear() on llquantize aggregation causes dtrace to
 exit OS-5501 printa() of multiple aggregations can fail for llquantize()
 Reviewed by: Patrick Mooney <patrick.mooney@joyent.com> Reviewed by: Robert
 Mustacchi <rm@joyent.com>

---
 .../test/tst/common/llquantize/tst.clear.d    | 23 ++++++++++++++++++
 .../tst/common/llquantize/tst.clear.d.out     |  6 +++++
 .../tst/common/llquantize/tst.multiaggs.d     | 24 +++++++++++++++++++
 .../tst/common/llquantize/tst.multiaggs.d.out | 13 ++++++++++
 usr/src/lib/libdtrace/common/dt_aggregate.c   | 23 +++++++++++-------
 usr/src/pkg/manifests/system-dtrace-tests.mf  |  4 ++++
 6 files changed, 85 insertions(+), 8 deletions(-)
 create mode 100644 usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d
 create mode 100644 usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d.out
 create mode 100644 usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d
 create mode 100644 usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d.out

diff --git a/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d
new file mode 100644
index 0000000000..6149546f61
--- /dev/null
+++ b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d
@@ -0,0 +1,23 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ */
+
+#pragma D option quiet
+
+BEGIN
+{
+	@ = llquantize(0, 10, 0, 6, 20);
+	clear(@);
+	exit(0);
+}
diff --git a/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d.out b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d.out
new file mode 100644
index 0000000000..34cfadafa5
--- /dev/null
+++ b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.clear.d.out
@@ -0,0 +1,6 @@
+
+
+           value  ------------- Distribution ------------- count    
+             < 1 |                                         0        
+               1 |                                         0        
+
diff --git a/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d
new file mode 100644
index 0000000000..bf2875d1f9
--- /dev/null
+++ b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d
@@ -0,0 +1,24 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ */
+
+#pragma D option quiet
+
+BEGIN
+{
+	@sfo["tabs"] = llquantize(10000, 10, 0, 6, 20);
+	@yvr["spaces"] = count();
+	printa(@sfo, @yvr);
+	exit(0);
+}
diff --git a/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d.out b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d.out
new file mode 100644
index 0000000000..139e0082ab
--- /dev/null
+++ b/usr/src/cmd/dtrace/test/tst/common/llquantize/tst.multiaggs.d.out
@@ -0,0 +1,13 @@
+
+  spaces                                            
+           value  ------------- Distribution ------------- count    
+             < 1 |                                         0        
+               1 |                                         0        
+                1
+  tabs                                              
+           value  ------------- Distribution ------------- count    
+            9500 |                                         0        
+           10000 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ 1        
+           15000 |                                         0        
+                0
+
diff --git a/usr/src/lib/libdtrace/common/dt_aggregate.c b/usr/src/lib/libdtrace/common/dt_aggregate.c
index 64ea79fbcc..a0eebe5a75 100644
--- a/usr/src/lib/libdtrace/common/dt_aggregate.c
+++ b/usr/src/lib/libdtrace/common/dt_aggregate.c
@@ -25,7 +25,7 @@
  */
 
 /*
- * Copyright (c) 2013, Joyent, Inc. All rights reserved.
+ * Copyright (c) 2016, Joyent, Inc. All rights reserved.
  * Copyright (c) 2012 by Delphix. All rights reserved.
  */
 
@@ -1145,7 +1145,13 @@ dt_aggwalk_rval(dtrace_hdl_t *dtp, dt_ahashent_t *h, int rval)
 		size = rec->dtrd_size;
 		data = &h->dtahe_data;
 
-		if (rec->dtrd_action == DTRACEAGG_LQUANTIZE) {
+		if (rec->dtrd_action == DTRACEAGG_LQUANTIZE ||
+		    rec->dtrd_action == DTRACEAGG_LLQUANTIZE) {
+			/*
+			 * For lquantize() and llquantize(), we want to be
+			 * sure to not zero the aggregation parameters; step
+			 * over them and adjust our size accordingly.
+			 */
 			offs = sizeof (uint64_t);
 			size -= sizeof (uint64_t);
 		}
@@ -1894,12 +1900,13 @@ dtrace_aggregate_walk_joined(dtrace_hdl_t *dtp, dtrace_aggvarid_t *aggvars,
 		rec = &aggdesc->dtagd_rec[aggdesc->dtagd_nrecs - 1];
 
 		/*
-		 * Now for the more complicated part.  If (and only if) this
-		 * is an lquantize() aggregating action, zero-filled data is
-		 * not equivalent to an empty record:  we must also get the
-		 * parameters for the lquantize().
+		 * Now for the more complicated part.  For the lquantize() and
+		 * llquantize() aggregating actions, zero-filled data is not
+		 * equivalent to an empty record:  we must also get the
+		 * parameters for the lquantize()/llquantize().
 		 */
-		if (rec->dtrd_action == DTRACEAGG_LQUANTIZE) {
+		if (rec->dtrd_action == DTRACEAGG_LQUANTIZE ||
+		    rec->dtrd_action == DTRACEAGG_LLQUANTIZE) {
 			if (aggdata->dtada_data != NULL) {
 				/*
 				 * The easier case here is if we actually have
@@ -1920,7 +1927,7 @@ dtrace_aggregate_walk_joined(dtrace_hdl_t *dtp, dtrace_aggvarid_t *aggvars,
 				 * -- either directly or indirectly.) So as
 				 * gross as it is, we'll grovel around in the
 				 * compiler-generated information to find the
-				 * lquantize() parameters.
+				 * lquantize()/llquantize() parameters.
 				 */
 				dtrace_stmtdesc_t *sdp;
 				dt_ident_t *aid;
diff --git a/usr/src/pkg/manifests/system-dtrace-tests.mf b/usr/src/pkg/manifests/system-dtrace-tests.mf
index 119a1ab070..5468c438b5 100644
--- a/usr/src/pkg/manifests/system-dtrace-tests.mf
+++ b/usr/src/pkg/manifests/system-dtrace-tests.mf
@@ -1038,6 +1038,10 @@ file path=opt/SUNWdtrt/tst/common/llquantize/tst.bases.d mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.bases.d.out mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.basic.d mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.basic.d.out mode=0444
+file path=opt/SUNWdtrt/tst/common/llquantize/tst.clear.d mode=0444
+file path=opt/SUNWdtrt/tst/common/llquantize/tst.clear.d.out mode=0444
+file path=opt/SUNWdtrt/tst/common/llquantize/tst.multiaggs.d mode=0444
+file path=opt/SUNWdtrt/tst/common/llquantize/tst.multiaggs.d.out mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.negorder.d mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.negorder.d.out mode=0444
 file path=opt/SUNWdtrt/tst/common/llquantize/tst.negvalue.d mode=0444
-- 
2.21.0

